Type stability checks

albangossard opened the pull request.

JordiBolibar reviewed (approved):
Great initiative. I guess there is no way to correctly capture the output of a function AND test it at the same time using `@inferred`, right?

JordiBolibar reviewed (changes requested):
Apparently there are some type instabilities in a couple of tests.

albangossard commented:
> Great initiative. I guess there is no way to correctly capture the output of a function AND test it at the same time using @inferred, right?

No there is no straightforward way of doing this. We could try to dump the output of `@code_warntype` and look for a pattern with characters in red or "Union" of something but in my opinion that wouldn't be super robust.

> Apparently there are some type instabilities in a couple of tests.

This is because I fixed type instabilities in https://github.com/ODINN-SciML/Sleipnir.jl/pull/105 and we need to make a new release of Sleipnir.

albangossard commented:
> Great initiative. I guess there is no way to correctly capture the output of a function AND test it at the same time using @inferred, right?

Oups sorry I didn't understand your question correctly in my first response. No the line on which you're inferring needs to be a function call. I tried to add a variable assignment but it's failing.

albangossard commented:
I found a way to check instabilities deeper than the returned type by using `JET.@test_opt`. I started to update the tests but obviously we have type instabilities basically everywhere. In the meantime, I'm turning the PRs into drafts again.

albangossard commented:
I implemented the tests with JET in Sleipnir, Muninn and Huginn. This allowed me to find several sources of instabilities that I fixed. The status on type stability is the following:
- https://github.com/ODINN-SciML/Sleipnir.jl/pull/105: The function we need to be type stable for the AD were fixed. `initialize_glaciers` is still unstable because of the inputs we read from disk but I don't think it's a big deal.
- https://github.com/ODINN-SciML/Muninn.jl/pull/26: `MB_timestep` is not type stable because of `RasterStack`.
- Huginn (this PR): There are many instabilities. I tried to improve stability where this was easy to fix but we need to refactor some of the structs and get rid of as many `Union` as we can.

Since there are many ongoing works and that I won't be able to pursue this work in the coming days, I suggest to merge these PRs and to continue working on that later on.
@facusapienza21 what do you think?

facusapienza21 commented:
@albangossard let's discuss about this in our next meeting. I saw the changes and there are okey, but I would like to thing if it is better to merge this or merge first my open PR.
